webui/react/src/omnibar/Omnibar.tsx

```diff
@@ -1,37 +1,59 @@
 import OmnibarNpm from 'omnibar';
-import React, { useCallback } from 'react';
+import React, { useCallback, useEffect } from 'react';
 
-import { StoreAction, useStoreDispatch } from 'contexts/Store';
+import { StoreAction, useStore, useStoreDispatch } from 'contexts/Store';
+import handleError from 'ErrorHandler';
 import * as Tree from 'omnibar/tree-extension/index';
 import TreeNode from 'omnibar/tree-extension/TreeNode';
 import { BaseNode } from 'omnibar/tree-extension/types';
 import { isTreeNode } from 'omnibar/tree-extension/utils';
 
 import css from './Omnibar.module.scss';
 
+/*
+Ideally we wouldn't need to access the element like this.
+A potential option is use the value prop in combinatio with encoding the tree path into
+the options returned by the tree extension.
+*/
+const omnibarInput = () => document.querySelector(
+  '#omnibar input[type="text"]',
+) as (HTMLInputElement | null);
+
 const Omnibar: React.FC = () => {
   const storeDispatch = useStoreDispatch();
+  const { ui } = useStore();
 
-  const hideBar = useCallback(
-    () => storeDispatch({ type: StoreAction.HideOmnibar }),
-    [ storeDispatch ],
-  );
+  const hideBar = useCallback(() => {
+    storeDispatch({ type: StoreAction.HideOmnibar });
+  }, [ storeDispatch ]);
+
+  const onAction = useCallback(async (item, query) => {
+    const input: HTMLInputElement|null = omnibarInput();
 
-  const onAction = useCallback((item, query) => {
-    /*
-    Ideally we wouldn't need to access the element like this.
-    A potential option is use the value prop in combinatio with encoding the tree path into
-    the options returned by the tree extension.
-    */
-    const input: HTMLInputElement|null = document.querySelector('#omnibar input[type="text"]');
     if (!input) return;
     if (isTreeNode(item)) {
-      return Tree.onAction(input, item, query);
+      try {
+        await Tree.onAction(input, item, query);
+        if (item.closeBar) {
+          hideBar();
+        }
+      } catch (e) {
+        handleError(e);
+      }
+    }
+  }, [ hideBar ]);
+
+  useEffect(() => {
+    const input: HTMLInputElement|null = omnibarInput();
+    if (ui.omnibar.isShowing) {
+      if (input) {
+        input.focus();
+      }
     }
-  }, []);
+  }, [ ui.omnibar.isShowing ]);
 
   return (
-    <div className={css.base}>
+    <div className={css.base} style={{ display: ui.omnibar.isShowing ? 'unset' : 'none' }}>
```

Comment:
**question:** I thought about this a bit more. Is there a reason why the omnibar state needs to be handled outside of the Omnibar? Can all the omnibar related state (such as `ui.omnibar.isShowing` be handled within `Omnibar.tsx` itself? Such that `App.tsx` can just simply embed `<Omnibar />` and be done.

If it's because of the App level key emitter, you can just create your own key listener inside `Omnibar`.

Reply:
valid point but since we're not adding to app.tsx here I'd like to leave this off.